[wip] kvserver: add avgProposalToLocalApplicationLatency

wenyihu6 opened the pull request:
**kvserver/closedts: add latency based closed timestamp policies**

This patch adds latency based closed timestamp policies to
ctpb.LatencyBasedRangeClosedTimestampPolicy.

Part of: #59680
Release note: none

---

**kvserver/closedts: auto-tune closed ts**

For global table ranges, we aim to ensure follower replicas can serve
present-time reads by calculating a target closed timestamp that leads the
current time (using the `LEAD_FOR_GLOBAL_READS` policy). The goal of this
estimation is to be confident that by that timestamp that the all follower
replicas should have received this closed timestamps and can serve present time
reads without redirecting to the leaseholders. Note that this is an estimated
goal, not the exact lead time. It is still possible that some follower nodes are
really behind or we fail to bump closed timestamps.

Calculation:

```
raft_propagation_time = max_network_rtt*1.5 + raft_overhead(20ms)
side_propagation_time = max_network_rtt*0.5 + side_transport_close_interval(200ms by default)
closed_ts_propagation = max(raft_propagation_time,side_propagation_time)
closed_ts_at_sender = now(sender’s current clock) + maxClockOffset(500ms by
default) + buffer (25ms) + closed_ts_propagation
```

Currently, `max_network_rtt` is hardcoded at 150ms, leading to a default 800ms lead.

Problems with the current implementation:

Tuning trade-offs: A higher lead timestamp forces leaseholders to wait longer
for the current time to catch up before returning to client, incurring a high
write latency. A lower lead timestamp may result in followers not fully
replicating the changes before returning control to the client, forcing follower
replicas to redirect query to the leaseholder and increase read latency.

Cluster wide setting: kv.closed_timestamp.lead_for_global_reads_override only
applies to the entire cluster, not per range. For geographically distant
replicas, a short lead time may not be enough for full application. However,
increasing the lead time cluster wide can harm write performance for replicas
that are close together and don’t require as much time to replicate.

This commit adds a latency tracker to the side transport senders.

Currently, nodes create a map of closed timestamps for each policy without
consulting leaseholders. Senders pass this map to leaseholders, who decide
whether to "bump" their closed timestamp and which policy to use. Senders then
send this information to receivers to opt in ranges and pick the closed
timestamps based on the policies ranges opt in.

Currently there are only two policies in the map side transport senders create.
We lack fine grained control into the closed timestamps based on the ranges. The
new design adds 16 more policies based on observed latency between the leaseholder
replica node and the furthest replica node.

Periodically, policy refresher iterates over all leaseholders of a node and
refreshes its cached closed timestamp policies. This policy is then used by
side transport and by raft side closed timestamp computation.

kv.closed_timestamp.lead_for_global_reads_auto_tune.enabled can now be used
to auto-tune the lead time that ranges global tables use to publish
close timestamps. The kv.closed_timestamp.lead_for_global_reads_override
cluster setting takes precedence over this one. If auto-tuning is disabled or
no data is observed, the system falls back to a hardcoded computed lead time.

Resolves: #59680
Release note: none

wenyihu6 committed:
kvserver/closedts: use max() instead of conditional checks

Previously, conditional check is used to find the maximum
between sideTransportPropTime and raftTransportPropTime.

This patch changes it to use the max builtin for better
readability.

Release note: none
Epic: none

wenyihu6 committed:
kvserver/closedts: move leadTargetOverride above for better readability

Previously, if a cluster defines `kv.closed_timestamp.lead_for_global_reads_override`,
the lead time calclulated at the sender will just be overriden in the end.

This patch moves the override part earlier in the logic and
breaks on an early return for better readability.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141838)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver: add kv.closed_timestamp.lead_for_global_reads_auto_tune_interval

This commit adds a new cluster setting,
`kv.closed_timestamp.lead_for_global_reads_auto_tune_interval` to control how
frequently the system measures latency between nodes. If non-zero, the observed
latency between nodes in the cluster would be used to dynamically auto-tune
the lead time for ranges using the LEAD_FOR_GLOBAL_READS policy.
The setting defaults to 5 minutes and can be disabled by setting to 0.
It only takes effect when `side_transport_interval` is non-zero.

Note that setting is currently unused. Future commits will populate and
integrate it.

Part of: #59680
Release note: none

wenyihu6 committed:
kvserver: replace RangeClosedTimestampPolicy with LatencyBasedRangeClosedTimestampPolicy

This commit replaces roachpb.RangeClosedTimestampPolicy with
ctpb.LatencyBasedRangeClosedTimestampPolicy. Currently, it maintains the existing
behavior. Future commits will introduce additional enum values to add locality
information into the side-transport closed timestamp estimation.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/closedts: add policy refresher

This commit adds policy refresher which refreshes the closed
timestamp policies of leaseholder ranges based on observed
network latency at the interval configured by
lead_for_global_reads_auto_tune_interval.

Part of: #59680
Release note: none

wenyihu6 committed:
kvserver: move closed ts lead time computation into a helper function

This patch extracts the calculation of closed timestamp lead time for ranges
configured with global reads policies to a helper function that takes in
networkRTT as a parameter. Currently, it always takes in
closedts.DefaultMaxNetworkRTT. But future commits will change it to take in
a dynamically observed network latency based on policies.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/closedts: clean up TargetForPolicy

This commit cleans up the closedts.TargetForPolicy logic a
bit to simplify future commits.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/closedts: change lastClosed to a map

This commit changes streamState shared between sender and
receiver from a fixed size slice to a map, similar to
trackedRange. Currently, it does not change any existing
behavior. But future commits use this to make sure only
lastClosed will be populated for enums with a corresponding
roachpb.RangeClosedTimestampPolicy.

Part of: cockroachdb#59680
Release note: none

wenyihu6 committed:
kvserver/closedts: add latency based closed timestamp policies

This patch adds latency based closed timestamp policies to
ctpb.LatencyBasedRangeClosedTimestampPolicy.

Part of: #59680
Release note: none

wenyihu6 committed:
kvserver/closedts: auto-tune closed ts

For global table ranges, we aim to ensure follower replicas can serve
present-time reads by calculating a target closed timestamp that leads the
current time (using the `LEAD_FOR_GLOBAL_READS` policy). The goal of this
estimation is to be confident that by that timestamp that the all follower
replicas should have received this closed timestamps and can serve present time
reads without redirecting to the leaseholders. Note that this is an estimated
goal, not the exact lead time. It is still possible that some follower nodes are
really behind or we fail to bump closed timestamps.

Calculation:

```
raft_propagation_time = max_network_rtt*1.5 + raft_overhead(20ms)
side_propagation_time = max_network_rtt*0.5 + side_transport_close_interval(200ms by default)
closed_ts_propagation = max(raft_propagation_time,side_propagation_time)
closed_ts_at_sender = now(sender’s current clock) + maxClockOffset(500ms by
default) + buffer (25ms) + closed_ts_propagation
```

Currently, `max_network_rtt` is hardcoded at 150ms, leading to a default 800ms lead.

Problems with the current implementation:

Tuning trade-offs: A higher lead timestamp forces leaseholders to wait longer
for the current time to catch up before returning to client, incurring a high
write latency. A lower lead timestamp may result in followers not fully
replicating the changes before returning control to the client, forcing follower
replicas to redirect query to the leaseholder and increase read latency.

Cluster wide setting: kv.closed_timestamp.lead_for_global_reads_override only
applies to the entire cluster, not per range. For geographically distant
replicas, a short lead time may not be enough for full application. However,
increasing the lead time cluster wide can harm write performance for replicas
that are close together and don’t require as much time to replicate.

This commit adds a latency tracker to the side transport senders.

Currently, nodes create a map of closed timestamps for each policy without
consulting leaseholders. Senders pass this map to leaseholders, who decide
whether to "bump" their closed timestamp and which policy to use. Senders then
send this information to receivers to opt in ranges and pick the closed
timestamps based on the policies ranges opt in.

Currently there are only two policies in the map side transport senders create.
We lack fine grained control into the closed timestamps based on the ranges. The
new design adds 16 more policies based on observed latency between the leaseholder
replica node and the furthest replica node.

Periodically, policy refresher iterates over all leaseholders of a node and
refreshes its cached closed timestamp policies. This policy is then used by
side transport and by raft side closed timestamp computation.

kv.closed_timestamp.lead_for_global_reads_auto_tune.enabled can now be used
to auto-tune the lead time that ranges global tables use to publish
close timestamps. The kv.closed_timestamp.lead_for_global_reads_override
cluster setting takes precedence over this one. If auto-tuning is disabled or
no data is observed, the system falls back to a hardcoded computed lead time.

Resolves: #59680
Release note: none

wenyihu6 commented:
Went with a different route.
